h/settings.py

```diff
@@ -13,54 +14,76 @@ class SettingError(Exception):
     pass
 
 
-class DeprecatedSetting(object):
-    """A wrapper for deprecated settings which emits appropriate warnings."""
-
-    def __init__(self, setting, message):
-        self.setting = setting
-        self.message = message
-
-        # Test seam
-        self.warn = log.warn
-
-    def __call__(self, environ):
-        result = self.setting(environ)
-        if result is not None:
-            self.warn(self.warning)
-        return result
-
-    @property
-    def warning(self):
-        return 'use of {s} is deprecated: {m}'.format(s=self.setting,
-                                                      m=self.message)
-
-
-class EnvSetting(object):
-
-    """An (optionally typed) setting read from an environment variable."""
-
-    def __init__(self, setting, varname, type=None):
-        self.setting = setting
-        self.varname = varname
-        if type is not None:
-            self.type = type
-        else:
-            self.type = str
-
-    def __call__(self, environ):
-        if self.varname in environ:
+class SettingsManager(object):
+    def __init__(self, settings=None, environ=None):
+        self.settings = settings or {}
+        self._environ = environ or os.environ
```

Comment:
There's a bug here which is my fault. For example here it should set `settings["foo"]` to `"DEFAULT"` but it doesn't:

```python
>>> environ = {}
>>> settings = {}
>>> sm = SettingsManager(settings, environ)
>>> sm.set("foo", "FOO", default="DEFAULT", required=True)
>>> settings
{}
```

It can also fail to raise `SettingsError` for missing required settings in some cases.

The problem is `settings = settings or {}`. If `settings` is `None` then it will create its own `settings` dictionary by default, which is what you intended. But if `settings` is an empty dictionary `{}` that is also falsey so it will also create its own dictionary instead of modifying the passed in one! And so the `settings` that the calling code passed in does not get modified.

The fix is:

```
def configure(environ=None, settings=None):
    if environ is None:
        environ = os.environ
    if settings is None:
        settings = {}
```

(Although personally I think `settings` should just be required and not have a default.)